### DISCUSSION TITLE
Enhancing input data integrity check

### DISCUSSION THREAD

**Discussion by dsdenes** [xyflow/xyflow#2766] [New Features]
Dear authors,

I am reaching out to inquire about best practices for checking the integrity of input data. Specifically, in our use case, we are manually converting data from another diagram library's model, and we have encountered some challenges in ensuring the correctness of the input data.

Currently, there is no way to confirm that the data will be correctly rendered in ReactFlow. Additionally, ReactFlow is quite permissive with its inputs and only emits developer errors, which can lead to hidden data integrity issues.

To address this, we would like to know if there are any utility functions or methods that can be used to check for integrity errors in the raw data (nodes, edges). Additionally, we would appreciate the ability to parametrize ReactFlow components in order to throw errors immediately, rather than just emitting console errors.

Thank you for your time and expertise on this matter.

**Comment by chrtze** [comment]
Hi, React Flow currently doesn't include utilities to validate a flow, except for the typescript definitions for the properties. Of course, this cannot prevent the flow from not rendering correctly. This is due to the fact that the node's data can be anything and we don't know how your custom nodes and handles look like. Changing it so that the console warnings are thrown errors would have side effects of breaking the app when you don't catch the errors properly and the warnings are depending on your implementation (maybe you forgot to add handles to a node, have hidden a handle with css, data format is invalid).

Writing a custom validator function specific to your data format would probably be the easiest option as you know best how your custom nodes and setup look like. Would be happy about ideas if this could be a generalized function that could be included in React Flow.

**Reply by bcakmakoglu** [reply]
Random thought but I assume that OP used `data` to describe the whole structure of a flow. So not just `node.data` or `edge.data` but the whole Array of nodes/edges, but I might be wrong of course, it's just how I understood it.
It should be possible to validate everything else *except* for `node.data` / `edge.data` (as those can be any arbitrary data that is unknown to React Flow), shouldn't it?

So given a user fetches some nodes/edges from a database and has some structure like this (simplified of course)
```ts
const nodes = [
   {
      id: '1',
      position: { x: 0, y: 0 },
      data: { label: 'Node 1' },
   }
  // ... some more nodes
]
```

It should be possible to validate that the given parameters, `id` and `position` are valid.

So assuming someone were to fetch nodes and it looked like this
```ts
const nodes = [
   {
      id: 1,
      position: { x: 0, y: 0 },
      data: { label: 'Node 1' },
   }
  // ... some more nodes
]
```

React Flow could tell that `id` is not a string and hence invalid.

Though I kinda feel like this might be a bit out of scope for the core library and could just be a separate package that provides this sort of validator that a structure could be run through before passing it to the React Flow component altogether.

Just my 2 cents though 🥸

**Comment by dsdenes** [comment]
Thank you @chrtze and @bcakmakoglu for your thoughts.

> This is due to the fact that the node's data can be anything and we don't know how your custom nodes and handles look like

Yes, the data property can be anything, but the source, target, sourceHandle, and targetHandle properties have a fixed structure. I believe that all cases where a console.warn is emitted during the render process could be checked by a validator. For example:

- The edge's target property points to an existing node
- The edge's targetHandle property points to an existing edge within the target node
- The sourceHandle property has the type of 'source'
- The targetHandle property has the type of 'target'
- The edge's target is not in the node itself (to prevent infinite loops)

(We're already using a validator like this, and I think most of it is business logic free, so could be useful for others too)

> Changing it so that the console warnings are thrown errors would have side effects of breaking the app when you don't catch the errors properly and the warnings are depending on your implementation (maybe you forgot to add handles to a node, have hidden a handle with css, data format is invalid)

Yes, it would be great if it would brake the app, because then we would discover the issue before our customers 😄.  However, if throwing an error is not possible, it would be good if the ReactFlow component could emit some structured errors on a callback, so we could report it to our logging system. For example:

`<FlowEditor onRenderError={(msg: string) => {}} />`

**Reply by moklick** [reply]
Hey @dsdenes 

thanks for your explanation. I like that idea and I think we can add something like that. `onError(code, message)`, `onRenderError(code, message)` other ideas @chrtze ?
